pennylane_qiskit/converter.py

```diff
@@ -36,17 +36,83 @@
+        2. the args passed when calling the qfunc, i.e. ``(qc, None, 0.35, 1.7, 0.2)``. These
+           are assigned to the Parameters as the are ordered on the Qiskit QuantumCircuit:
+           alphabetized by Parameter name
+        3. Some combination of args and kwargs, i.e. ``(qc, None, 0.35, 0.2, psi=1.7)``
+        4. (legacy) ``params`` from the kwarg ``params`` of the qfunc call, which is expected
+            to already be a dictionary of the format ``{Parameter("name"): value}``, i.e.
+            ``(qc, {Parameter("phi"): 0.35, Parameter("psi"): 1.7, Parameter("theta"): 0.2})``
+        5. (legacy) ``params`` passed as a single arg, which is expected
+            to already be a dictionary of the format ``{Parameter("name"): value}``, i.e.
+            ``(qc, None, {Parameter("phi"): 0.35, Parameter("psi"): 1.7, Parameter("theta"): 0.2})``
+
+    Returns:
+        params (dict): A dictionary mapping ``quantum_circuit.parameters`` to values
+    """
+
+    # if no kwargs are passed, and a dictionary has been passed as a single argument, then assume it is params
+    if params is None and not kwargs and (len(args) == 1 and isinstance(args[0], dict)):
+        return args[0]
+
+    # make params dict if using args and/or kwargs
+    if args or kwargs:
+        if params is not None:
+            raise RuntimeError(
+                "Cannot define params via the params kwarg when passing Parameter values "
+                "as individual args or kwargs."
+            )
+
+        # create en empty params dict
+        params = {}
+
+        # populate it with any parameters defined as kwargs
+        for k, v in kwargs.items():
+            # the key needs to be the actual Parameter, whereas kwargs keys are parameter names
+            qc_param = [p for p in quantum_circuit.parameters if p.name == k]
+            if not qc_param:
+                param_names = ", ".join([p.name for p in quantum_circuit.parameters])
+                raise TypeError(
+                    f"Got unexpected parameter keyword argument '{k}'. Circuit contains parameters: {param_names}"
+                )
+            params[qc_param[0]] = v
+
+        # get any parameters not defined in kwargs (may be all of them) and match to args in order
+        arg_parameters = [p for p in quantum_circuit.parameters if p.name not in kwargs.keys()]
+        # if too many args were passed to the function call, raise an error
+        # all other checks regarding correct arguments will be processed in _check_circuit_and_assign_parameters
+        # (based on the full params dict generated by this function), but this information can only be captured here
+        if len(args) > len(arg_parameters):
+            raise TypeError(
+                f"Expected {len(arg_parameters)} positional argument{'s' if len(arg_parameters) > 1 else ''} but {len(args)} were given"
+            )
+        params.update(dict(zip(arg_parameters, args)))
```

Comment:
Running pylint on this file gives:
```
pennylane_qiskit/converter.py:103:81: C0201: Consider iterating the dictionary directly instead of calling .keys() (consider-iterating-dictionary)
pennylane_qiskit/converter.py:333:8: R1702: Too many nested blocks (6/5) (too-many-nested-blocks)
pennylane_qiskit/converter.py:249:4: R0912: Too many branches (19/12) (too-many-branches)
```

So basically this is becoming too nested and branched. Basically it's time to start pulling out helper functions.

Reply:
So pylint and sourcery are definitely complaining about this file, but they don't seem to be concerned about this function. And it doesn't seem particularly nested or branchy to me... 🤔

Reply:
I wonder if its placing the warning in the wrong place, since `_function` is overly nested, branchy, and overall low code quality.

Reply:
I'm going to ignore it for now, looks like the CodeFactor check still passes, and there are quite a few changes to this function in the next PR - I don't really want to create a bunch of merge conflicts by moving things around here.